### PR TITLE
[lora] refactor - native LoRA

### PR DESCRIPTION
**Pull request by yushengsu-thu**
Refactor PR **into #1792's branch** (`zhichen/lora-native`), so the native-LoRA extraction ships as a plugin from the start. Adapter math is unchanged — a side-by-side harness against the base monolith shows bit-identical adapter parameters, forward outputs, backward gradients, and HF exports for full-target GQA (TP=1, TP=2 ± SP) and MLA configs. **Bridge-mode LoRA is untouched.**

Includes #2072 (function-level redesign + fixes from validating four models end-to-end), the `lora-native` CI suite (#2073), and — after merging main's Inkling LoRA (#2122) — a rewrite of Inkling onto the plugin architecture, replacing the model-specific monolith.

## Layout

```python
miles_plugins/lora/
    __init__.py
    lora.py              # Single public entry point and lifecycle orchestration
    config.py            # LoRAConfig: rank, alpha, dropout, and target modules
    registry.py          # model_type -> LoRAArchSpec; launcher helpers
    hf_adapter.py        # HF/PEFT naming (incl. megatron->HF conversion), import, export
    sglang_adapter.py    # SGLang serving export (fused-buffer zero-fill, target expansion)
    checkpointing.py     # Adapter checkpoint state dicts and per-rank shard naming
    distributed.py       # TP/EP/SP communication and gradient handling (ParallelGather)

    spec/
      base.py            # Contracts: ProjectionSpec, AttachContext, protocols, enums
      layout.py          # Declarative layouts: dim resolvers, bindings, attach walk, spec bases
      attention.py       # GQA / MLA / GDN / hybrid attention spec classes
      mlp.py             # Fused gate/up + down spec class
      moe.py             # Routed-expert per-layer policy; Inkling grouped experts
      lm_head.py         # Output-head specs (Inkling's muP lm_head)

    modules/
      linear.py          # LoRALinear + LoRASplitAdapter (SplitQKV / SplitFC1)
      moe.py             # Grouped-GEMM expert, shared-experts, and lm_head adapters

```

## Design

**Classes carry the taxonomy; every fact is declared once and derived everywhere else.**

- **Declarative layouts** (`spec/layout.py`): an architecture spec is a class holding a `ModuleLayout` table — which projections exist, on which physical linear, with which shard geometry. One shared `attach_layout()` implements the existence-check → target-filter → guard → build → hook walk; the dim-resolver section is the only spec-side reader of MCore attribute names. A new layout is ~15 lines of facts.
- **Spec hierarchy** (`LayoutSpec` → `AttentionSpecBase` → `GQA/MLA/GDN/Inkling`, `HybridGQAGDN(GQAAttentionSpec)`, `SharedOuterExpertMoESpec(GeneralExpertMoESpec)`): `supported_targets`, the canonical `--target-modules` order, and SGLang's fused-family expansion all derive from the layout declaration; a new architecture's fused groups reach serving-side target expansion automatically. `ShardLayout`/`AttentionFamily` str-enums replace bare strings; no parallel name tables in the exporters.
- **Public export descriptors**: adapters expose `exports() -> ProjectionExport` (with `SGLangFusedGroup` metadata); `sglang_adapter.py` consumes only that API.
- **Adapter IO hooks**: modules whose export shape isn't a per-projection column/row gather override `export_plan(gather)` / `load_plan_custom(take)`; `hf_adapter.py` stays generic. `ParallelGather` batches async all-gathers over the TP *or* EP group behind one `request(tensor, dim, group=...)` call.
- **Escape hatches for non-table structure**: `ModuleLayout.hf_block_prefix` and `ProjectionBinding.adapter_class` override HF naming and the module class per binding; `LoRAArchSpec` gained `moe.attach(...)` and an optional `lm_head` spec so a fully custom family still routes through the one orchestrator in `lora.py`.
- **Launcher helpers**: `default_target_modules(hf_checkpoint)` and `preflight_native_lora(...)` (registry ∧ mbridge bridge ∧ model-args audit, no GPU) — launch scripts stop re-declaring per-family target strings and fail in seconds when a conversion bridge is missing.

## Inkling on the plugin (replaces the #2122 monolith)

`miles_plugins/models/inkling/lora.py` (759 lines, its own attach/export/IO stack) is deleted; Inkling's specs live in the role files next to the base classes they inherit from (`InklingAttentionSpec` in `spec/attention.py`, `InklingDenseMLPSpec` in `spec/mlp.py`, `InklingMoESpec` in `spec/moe.py`, `InklingLMHeadSpec` in `spec/lm_head.py`), registered as `inkling_mm_model`:

- **Attention**: 4-way fused `[q;k;v;r]` declared as one `FusedAttach` — Inkling's fusion is a plain concat (no GQA group-permute), so the split-adapter row table `{q,k,v,r}` is the entire model-specific code. TML export names (`attn.wq_du/wk_dv/wv_dv/wr_du/wo_ud`) come straight from the `ProjectionSpec`s.
- **Dense MLP**: fused gate/up exports as the single TML `gate_up_proj` projection via an `adapter_class` override whose `export_plan` gathers the gate/up halves separately (a plain dim-0 TP gather would interleave ranks).
- **Routed experts**: `LoRAGroupedFC1/FC2` — shared-A, per-expert-B grouped-GEMM deltas (`F.grouped_mm` over the router's token offsets), EP-aware export/load. **Shared experts**: one adapter with per-sub-expert B slices, hooked into each sub-expert's forward.
- **lm_head**: `LoRAOutputHead` with muP logit scaling and unpadded-vocab trim.
- **Serving**: TML names are auto-detected by SGLang, so the spec declares `sglang_lora_target_modules=("all",)` and `sglang_adapter.sglang_target_modules()` answers launch/sync target lists for every native run — core carries no model-specific serving logic.
- `normalize_targets` maps any request (`all-linear` or HF-style names) to the full native TML set, matching the provider it replaces.

## Fixes found by running the models natively

- **`miles_plugins/mbridge/kimi_k25.py`** (new): raw-mode K2.5 conversion needs an mbridge bridge (bridge mode uses megatron.bridge's own `KimiK25VL` bridge and never converts). Thin `DeepseekV3Bridge` subclass: `text_config` descent + `language_model.` prefix remap.
- **`megatron_to_hf` kimi dispatch**: `_convert_to_hf_core` matched only `kimi_k25`, but the direct weight iterator passes `kimik25config` (dead branch); the converter also lacked the raw text-only provider's unprefixed top-level weight names.
- **`convert_kimi_int4_to_bf16.py` strips `quantization_config`**: a surviving one sends SGLang through its CompressedTensors path, which serves the BF16 checkpoint with a degenerate context-free forward → garbage rollouts, train/rollout `logprob_abs_diff` ~2.2. Verified by a megatron-vs-HF-vs-SGLang three-way logprob comparison.
- **`torch_dsa_topk` clamps k to the key length**: short sequences (< `index_topk` keys) crashed `torch.topk` in raw-mode log-prob compute.
- **`run_glm5_2_744b_a40b_lora_native.py`** (new): raw-mode GLM-5.2 launcher; its toy conversion must be single-rank (any PP split of the 5-layer toy starts a stage on a DSA skip layer).

## CI

`tests/e2e/lora_native/` (label `lora-native`, trigger `run-ci-lora-native`): qwen3, qwen3.5, glm4.7, glm5.2-5layer, kimi-2.5-2layer, and inkling-small-4layer — native runs checking logprob parity and weight-update equality (vision towers and engine-transformed params skip-listed).

## Validation

20-rollout native GRPO runs (wandb `ch271828n-team/miles_lora_native`), acceptance ≈ 0.01 train/rollout `logprob_abs_diff`:

| model | logprob_abs_diff |
|---|---|
| Qwen3-8B | 0.0099 |
| Qwen3.5-35B-A3B | 0.0105 (grad_norm ~1e-2 throughout; the previously recorded GDN raw-backward divergence did not reproduce) |
| Qwen3.5-9B (dense hybrid GDN) | 0.0045–0.0090 (mean 0.0063), grad_norm ≤ 0.032 — divergence absent on the dense variant too |
| GLM-5.2 5-layer | 0.0096 |
| Kimi-K2.5 2-layer | 0.0124 (after the quantization_config fix) |
| Inkling-Small (full 42-layer, 4 nodes / 32×H200, TP4·EP4·PP8, 16-GPU engines) | 0.0091–0.0123 (mean 0.0107) over 10 rollouts; reward stable ~0.5 on dapo-math |

The Inkling run exercises the whole custom stack at multi-node scale: 4-way fused [q;k;v;r] export, grouped-GEMM expert adapters under EP4, shared-experts hooks, the muP lm_head, and SGLang serving via the `["all"]` sentinel. (SGLang note: LoRA-on-MoE serving currently supports only the triton MoE runner — `flashinfer_trtllm_routed` fails engine init with NotImplementedError.)

`tests/fast/miles_plugins/lora`: 64 passed (includes Inkling spec locks). Behavior-preservation smoke: a 2-rollout Qwen3-8B run reproduces the baseline's first-rollout logprob_abs_diff (0.00967 vs 0.00978). Adapter attribute names are pinned, so checkpoint keys are unchanged.

### CODE REVIEW

File: miles/backends/megatron_utils/lora_utils.py [hunk 1]
```diff
@@ -408,6 +408,24 @@ def create_lora_instance(args: Namespace):
     return lora
 
 
+_DEFAULT_LORA_PROVIDER = "miles_plugins.lora.lora"
```
Comment (yushengsu-thu) [inline miles/backends/megatron_utils/lora_utils.py]: move to the top because it's the gobal var

File: miles_plugins/lora/distributed.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+   need no adapter-specific overlap handling; for grouped-expert LoRA bridge even
+   wraps the 3D weight in a callable module so the pre-hook that drives
+   overlapped param gather fires (``peft/utils.py:1706-1720``). Native attaches
+   siblings and monkeypatches the wrapped module's forward, which forces the
+   out-of-band cross-TP grad sum below over the same ``main_grad`` buffer DDP
+   reduce-scatters — hence the ``--overlap-grad-reduce`` /
+   ``--overlap-param-gather`` refusals in ``lora.py``. Lifting them means making
+   the adapter a real parallel linear (or at least a callable weight container),
+   not adding more collectives here.
+"""
```
Comment (yushengsu-thu) [inline miles_plugins/lora/distributed.py]: only write the abstruct here and list the to-do and unsupport parts

File: miles_plugins/lora/spec/moe.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+   ``(probs, routing_map)`` and interposes jitter / float32 expert bias /
+   load balancing between gating and routing, so the delta must be injected
+   mid-forward. Note SGLang has no router-LoRA buffer type, so it would train
+   but not serve.
+
+Related roadmap notes live next to the code that would change: the expert-TP
+group in ``miles_plugins/lora/distributed.py``, the expert-axis export in
+``miles_plugins/lora/codec/hf.py``, and GDN / uncompressed-MLA attention in
+``miles_plugins/lora/spec/attention.py``.
+"""
```
Comment (yushengsu-thu) [inline miles_plugins/lora/spec/moe.py]: only list the abstruct here and list the to-do/unsupported parts

**Review by yushengsu-thu — COMMENTED** [review]
fix them

**Comment by yushengsu-thu** [comment]
To-do

- dsk-v4 support: because we have not support dsk-v4 lora on sglang side yet. Thus, I reomve dsk-v4 part and we can add it after sglang side support.

- Refer to megatron-bridge to compelete the features at below layouts:

- - [miles_plugins/lora/spec/moe.py](https://github.com/radixark/miles/pull/2017/files/173ca6da5a2d65bbf171afb2bcf222fac7197c2e#diff-b337c1cfee6697c1a224775a17c8e57160ca1b85589382934b1b61c35fad05c1)

- - [miles_plugins/lora/distributed.py](https://github.com/radixark/miles/pull/2017/files/173ca6da5a2d65bbf171afb2bcf222fac7197c2e#diff-583815cfa39ffbc5b7a0bead06c00f273dc5d039c8adb95b60ed04bcdc4bbe37)


- [I added] ~~moe lora: in native branch, it only supports shared moe lora (bth, that's tml special design moe lora). Another general moe lora is non-shared and this is implemented in sglang and megatron-bridge already as well. We can refer to that part when we refactor megatron-bridge lora in miles. (Refer to different type of lora in megatron-bridge).~~

File: miles_plugins/lora/distributed.py (hunk 2 above)
Comment (yushengsu-thu) [reply miles_plugins/lora/distributed.py]: Done — though you had already rewritten this one on the branch, so I kept your version and applied the same shape everywhere else it was still missing: `codec/hf.py`, `codec/checkpoint.py`, `modules/linear.py`, `modules/moe.py`, `spec/base.py`, `GDNAttentionSpec`/`MLAAttentionSpec`, the plugin `__init__`, several `lora_utils.py` docstrings, and the docs parallelism section (489b015).

Assertion and warning text is deliberately left long — those are user-facing diagnostics where naming the reason and the escape hatch is the point.

File: miles/backends/megatron_utils/lora_utils.py (hunk 1 above)
Comment (yushengsu-thu) [reply miles/backends/megatron_utils/lora_utils.py]: Already handled in 4424556 — `_DEFAULT_LORA_PROVIDER` is at module scope now (`lora_utils.py:18`), alongside `_NATIVE_LORA_PROVIDER_PATHS`.

File: miles_plugins/lora/spec/moe.py (hunk 3 above)
Comment (yushengsu-thu) [reply miles_plugins/lora/spec/moe.py]: Done — your rewrite of this file is what shipped; I matched its shape in the remaining files (489b015).

File: miles_plugins/lora/checkpointing.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+
+import torch
+import torch.nn as nn
+
+from miles_plugins.lora.modules.linear import iter_adapters
+
+_MODEL_CHUNK_PREFIX = "_miles_model_chunks."
```
Comment (Zhichenzzz) [inline miles_plugins/lora/checkpointing.py]: here we don't need to make the `miles-sepcific` lora weights, how about aligning with the megatron-bridge naming ways?

Comment (yushengsu-thu) [reply miles_plugins/lora/checkpointing.py]: use mcore checkpoint and sharding

File: miles/backends/megatron_utils/megatron_to_hf/__init__.py [hunk 5]
```diff
@@ -58,7 +58,9 @@ def _convert_to_hf_core(args, model_name, name, param):
         converted_named_tensors = convert_mimo_to_hf(args, name, param)
     elif "kimivl" in model_name:
         converted_named_tensors = convert_kimivl_to_hf(args, name, param)
-    elif "kimi_k25" in model_name:
+    elif "kimi_k25" in model_name or "kimik25" in model_name:
+        # KimiK25Config lowercases to "kimik25config"; the underscore spelling
+        # covers callers that pass the HF model_type instead.
```
Comment (yushengsu-thu) [inline miles/backends/megatron_utils/megatron_to_hf/__init__.py]: remove

File: miles_plugins/lora/checkpointing.py [hunk 6]
```diff
@@ -1,29 +1,237 @@
...
+def _sharded_entry_dict_key(entry: Any) -> str:
+    """A layout-invariant, per-shard-unique dict key for one ShardedTensor.
+
+    Plain state-dict keys cannot be used to merge model chunks: MCore keys the
+    walk's dict by chunk-LOCAL layer indices (only the ShardedTensor ``.key``
+    is globalized), so virtual-pipeline chunks collide on identical dict keys.
+    ``.key`` alone is not unique either — with homogeneous-layer checkpoints
+    every layer shares one ``.key`` and is distinguished by its prepended
+    global offset. The pair is unique.
+    """
```
Comment (yushengsu-thu) [inline miles_plugins/lora/checkpointing.py]: make it shorter

File: miles_plugins/lora/checkpointing.py [hunk 7]
```diff
@@ -1,29 +1,237 @@
...
+    Walks each chunk's standard MCore ``sharded_state_dict`` with the opt-in
+    metadata flag so ``NativeLoRAAdapter`` emits real ShardedTensors — global
+    layer keys, PP offsets, and TP axes all come from the standard walk — then
+    filters to adapter parameters by object identity and re-keys the merged
+    dict by ``(ShardedTensor.key, global_offset)`` (see
+    ``_sharded_entry_dict_key``). Fails loudly if any attached adapter
+    parameter never surfaced: a provider whose module tree does not recurse
+    into adapter children would otherwise silently drop weights from every
+    checkpoint.
+    """
```
Comment (yushengsu-thu) [inline miles_plugins/lora/checkpointing.py:79]: make it shorter

File: miles_plugins/lora/checkpointing.py [hunk 8]
```diff
@@ -1,29 +1,237 @@
...
+    optimizer: Any | None = None,
+    opt_param_scheduler: Any | None = None,
+    iteration: int | None = None,
+) -> None:
+    """Write adapters (+ optimizer/scheduler state) as one MCore dist checkpoint.
+
+    Collective across the default process group — every rank must call it.
+    Scheduler state and the iteration are rank-identical and land in the
+    checkpoint's ``common.pt``.
+    """
```
Comment (yushengsu-thu) [inline miles_plugins/lora/checkpointing.py:114]: make it shorter

File: miles_plugins/lora/checkpointing.py [hunk 9]
```diff
@@ -1,29 +1,237 @@
...
+    checkpoint (e.g. a different ``--target-modules`` set) raise inside
+    ``dist_checkpointing.load`` rather than silently reinitializing; the
+    checkpoint is one collectively-validated unit, so there is no
+    per-rank-file consensus step as in the legacy format.
+
+    Returns the saved iteration when optimizer state was restored, else None
+    (matching the legacy resume contract: no optimizer state means training
+    restarts its schedule, keeping fp32 main params aligned via
+    ``reload_model_params``).
+    """
```
Comment (yushengsu-thu) [inline miles_plugins/lora/checkpointing.py:157]: make it shorter

File: miles_plugins/lora/checkpointing.py [hunk 10]
```diff
@@ -51,67 +259,54 @@ def apply(self) -> int:
...
+    """Preflight a legacy adapter state dict without mutating parameters.
+
+    Two legacy key schemes are readable. Shards from earlier revisions of this
+    branch qualified every key as ``_miles_model_chunks.{chunk}.{name}`` and
+    map back exactly. Older flat shards keyed parameters by their unqualified
+    chunk-local names; a flat key is accepted whenever it identifies exactly
+    one parameter across all current chunks, and a key shared by multiple
+    virtual-pipeline chunks is rejected as ambiguous — flat writers overwrote
+    those values and there is no safe way to reconstruct which chunk the
+    surviving tensor came from.
```
Comment (yushengsu-thu) [inline miles_plugins/lora/checkpointing.py]: make it shorter

**Review by yushengsu-thu — COMMENTED** [review]
clean here
